lib/AtomicCache.php

```diff
@@ -0,0 +1,176 @@
+            $value = yield $this->cache->get($key);
+
+            if ($value !== null) {
+                return $value;
+            }
+
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            try {
+                // Attempt to get the value again, since it may have been set while obtaining the lock.
+                $value = yield $this->cache->get($key);
+
+                if ($value !== null) {
+                    return $value;
+                }
+
+                return yield from $this->create($create, $key, null, $ttl);
+            } finally {
+                $lock->release();
+            }
+        });
+    }
+
+    /**
+     * The key is locked, the current value in the cache is accessed, then the $create callback is invoked with the key
+     * as the first parameter and the current value as the second parameter. The value returned from the callback is
+     * stored in cache and the promise returned from this method is resolved with the value.
+     *
+     * @param string   $key
+     * @param callable(string $key, mixed $value) $create
+     * @param int|null $ttl
+     *
+     * @return Promise<string>
+     *
+     * @throws CacheException If the $create callback throws an exception while generating the value.
+     */
+    public function swap(string $key, callable $create, ?int $ttl = null): Promise
+    {
+        return call(function () use ($key, $create, $ttl): \Generator {
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            $value = yield $this->cache->get($key);
+
+            try {
+                return yield from $this->create($create, $key, $value, $ttl);
+            } finally {
+                $lock->release();
+            }
+        });
+    }
+
+    private function lock(string $key): \Generator
+    {
+        try {
+            return yield $this->mutex->acquire($key);
+        } catch (\Throwable $exception) {
+            throw new CacheException(
+                \sprintf('Exception thrown when obtaining the lock for key "%s"', $key),
+                0,
+                $exception
+            );
+        }
+    }
+
+    private function create(callable $create, string $key, ?string $value, ?int $ttl): \Generator
+    {
+        try {
+            $value = yield call($create, $key, $value);
+        } catch (\Throwable $exception) {
+            throw new CacheException(
+                \sprintf('Exception thrown while creating the value for key "%s"', $key),
+                0,
+                $exception
+            );
+        }
+
+        if (!\is_string($value)) {
+            throw new CacheException(\sprintf(
+                'The value to be cached for key "%s" must be a string, %s returned',
+                $key,
+                \is_object($value) ? \get_class($value) : \gettype($value)
+            ));
+        }
```

Comment:
I think we should accept a [serializer like in `http-server-session`](https://github.com/amphp/http-server-session/blob/1178eeabe0c12e5993fc824db0d6536a15a753c4/src/Serializer.php), except that the serializer should accept any object instead of just arrays like in `http-server-session`.

Reply:
I just added such a [`Serializer` interface in parallel](https://github.com/amphp/parallel/blob/f5cb1e87e8f6540ce4eb8e32743003c2003b5c4f/lib/Sync/Serializer.php). It hasn't been in a tagged version yet. Let's move it to sync so it can be reused in this package and parallel.

Reply:
@trowski Why to `sync`? Maybe directly to Amp itself?

Reply:
Because it's related to data sharing, which libs doing such things require `sync`. If you want to put it in Amp that's fine by me, the interface and single implementation are pretty simple. I'd keep more complex implementations like the one used in `http-server-session` in that library.

Reply:
Is there a reason for `http-server-session` to use different `Serializer` interface? Couldn't is use the same as `parallel` and `cache`?

Reply:
@enumag Yes, `http-server-session` would use the same interface. There's a compressing serializer in that library I was thinking should stay there, but that implementation is pretty simple too, so I guess it could be moved as well.